### DISCUSSION TITLE
Terraform modules for Talos OS

### DISCUSSION THREAD

**Discussion by ilvpozzd** [siderolabs/talos#5606] [Show and tell]
I have made Terraform modules that fully describe the Talos OS 1.0.x configuration and deploy virtual machines as a single cluster in the vSphere corporate cloud.

Cluster Module - https://github.com/ilpozzd/terraform-talos-vsphere-cluster
Virtual Machine Module - https://github.com/ilpozzd/terraform-talos-vsphere-vm
Terraform-managed secrets module (generates all secret data as certs, ids, hashes e.g.) - https://github.com/ilpozzd/terraform-talos-secrets

All modules also available in Terraform Registry

And Terragrunt example to multi-region infrastructure - https://github.com/ilpozzd/talos-vsphere-cluster-terragrunt-example

**Comment by smira** [comment]
thanks, Ilya!

I noticed that secrets module is generating secrets without Talos machinery package (see Example here: https://pkg.go.dev/github.com/talos-systems/talos/pkg/machinery@v1.0.5/config/types/v1alpha1/generate). I wonde if that is something hard to integrate with Terraform?

**Comment by ilvpozzd** [comment]
@smira, at the moment, I know two ways of integration - this is to use directly talosctl from the Terraform module, but this will require a user-installed utility and a large number of operations with the file system (I left this approach in favor of terraform-managed resources). The second option is to write a Terraform provider that works with the Go talosctl library or does similar things without it.

**Reply by smira** [reply]
I think using Talos Go library (machinery package) is the only way which will guarantee future compatibility. As we keep adding features to Talos, base machine config for the next version might have some new features.

Talos machine config is backwards compatible though - config for v1.0 will work with v1.1 without any changes.

Machinery package also provides machine config generation for previous versions of Talos, so you can use machinery package v1.1 to generate machine config compatible with v1.0.

**Reply by ilvpozzd** [reply]
Yes, I think this is the best option. Terraform Registry already has several providers that implement the functionality. Modules are more suited to specific solutions (as is the case with vSphere). I will continue to update their config to match the Talos config.

**Comment by ilvpozzd** [comment]
I have allocated the userdata generation module into a separate module. Now you can use the resulting configuration in any cloud that supports VM initialization via user-data!

https://registry.terraform.io/modules/ilpozzd/vm-userdata/talos/latest

Also updated all modules to a new version, which also uses a separate module for user-data
